### PR TITLE
276 add patterns to enemies

### PR DESCRIPTION
**Pull request by 31Nathan**
This PR will implement patterns for enemies corresponding to their types
Each pattern is described in the following file: EnemyPatterns.cpp

### CODE REVIEW

File: libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    //     }
+    //     if (x + 2 < nbEnemies) {
+    //         if (firstHasBeenModified) {
+    //             applySquarePatterns(iceEnemies.at(x + 3), iceEnemies.at(x + 2));
+    //         } else {
+    //             applySquarePatterns(iceEnemies.at(x), iceEnemies.at(x + 2));
+    //             firstHasBeenModified = true;
+    //         }
+    //     }
+    //     if (x + 1 < nbEnemies) {
```
Comment (31Nathan) [inline libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp:211]: @osvegn

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemy.cpp [hunk 2]
```diff
@@ -26,16 +26,16 @@ namespace ecs
             .addComponent<Collidable>()
             .addComponent<Velocity>(multiplierAbscissa, multiplierOrdinate)
             .addComponent<Enemy>(type);
+            RandomDevice &random = world.getResource<RandomDevice>();
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemy.cpp:29]: Check discord

File: libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+/// @brief Apply the electric Pattern (flash shape) to the enemy
+/// @param world the world where the super fast enemy is
+/// @param enemy the super fast enemy
+static void electricEnemyPatterns(World &world, std::shared_ptr<Entity> enemy)
+{
+    makeEnemyGoRandom(world, enemy, std::pair<int, int>(100, 1800), std::pair<int, int>(100, 900));
+}
+
+/// @brief  Init the Ice Pattern (hord squared shape pattern)
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp:100]: Double space

File: libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                std::pair<int, int>(baseSquareX, baseSquareY - SQUARE_SPACE));
+        }
+        if (x + 3 < nbEnemies) {
+            setIceEnemyBaseParam(iceEnemies.at(x + 3), std::pair<int, int>(baseSquareX, baseSquareY - SQUARE_SPACE),
+                std::pair<int, int>(baseSquareX, baseSquareY));
+        }
+    }
+}
+
+/// @brief  Apply the Ice Pattern (hord squared shape pattern)
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp:148]: Double space

File: libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+            auto secondGuard = std::lock_guard(*previousEnemy.get());
+            Position &newDest = previousEnemy.get()->getComponent<Position>();
+
+            dest.x = newDest.x;
+            dest.y = newDest.y;
+            vel.multiplierAbscissa = 0;
+            vel.multiplierOrdinate = 0;
+            vel.modified = true;
+            dest.modified = true;
+
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp:198]: Useless empty line

File: libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+            applySquarePatterns(iceEnemies.at(x + 2), iceEnemies.at(x + 1));
+            applySquarePatterns(iceEnemies.at(x + 1), iceEnemies.at(x));
+            continue;
+        }
+        if (x + 1 < nbEnemies) {
+            applySquarePatterns(iceEnemies.at(x), iceEnemies.at(x + 1));
+            applySquarePatterns(iceEnemies.at(x + 1), iceEnemies.at(x));
+            continue;
+        }
+        {
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp:225]: What ????

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
Some doc, typo and question, otherwise Masterclass

File: libs/R-TypeLogic/Global/Systems/ElectricInvisibleEnemySystem.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+** ElectricInvisibleEnemy
+*/
+
+#include "ElectricInvisibleEnemySystem.hpp"
+#include "R-TypeLogic/Global/Components/InvisibleComponent.hpp"
+#include "R-TypeLogic/Global/Components/EnemyComponent.hpp"
+#include "R-TypeLogic/Global/Components/PositionComponent.hpp"
+#include "R-TypeLogic/Global/Components/DestinationComponent.hpp"
+
+using namespace ecs;
```
Comment (TomDUVAL-MAHE) [inline libs/R-TypeLogic/Global/Systems/ElectricInvisibleEnemySystem.cpp:14]: It's not R-TypeLogic::Global::Systems ?

File: libs/R-TypeLogic/Global/Systems/ElectricInvisibleEnemySystem.hpp [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+** ElectricInvisibleEnemy
+*/
+
+#ifndef ELECTRICINVISIBLEENEMY_HPP_
+#define ELECTRICINVISIBLEENEMY_HPP_
+
+#include "World/World.hpp"
+#include "R-TypeLogic/Global/Components/InvisibleComponent.hpp"
+
+namespace ecs
```
Comment (TomDUVAL-MAHE) [inline libs/R-TypeLogic/Global/Systems/ElectricInvisibleEnemySystem.hpp:14]: It's not R-TypeLogic::Global::Systems ?

File: libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+#include "R-TypeLogic/Global/Components/VelocityComponent.hpp"
+#include "R-TypeLogic/Global/SharedResources/Random.hpp"
+
+#define MINIMUM_WIDTH  1200
+#define MAXIMUM_WIDTH  1500
+#define MINIMUM_HEIGTH 400
+#define MAXIMUM_HEIGTH 900
+#define SQUARE_SPACE   300
+
+using namespace ecs;
```
Comment (TomDUVAL-MAHE) [inline libs/R-TypeLogic/Server/Systems/EnemiesPatterns.cpp:24]: It's not R-TypeLogic::Server::Systems ?

File: libs/R-TypeLogic/Server/Systems/EnemiesPatterns.hpp [hunk 10]
```diff
@@ -16,15 +16,15 @@
 
 namespace ecs
```
Comment (TomDUVAL-MAHE) [inline libs/R-TypeLogic/Server/Systems/EnemiesPatterns.hpp:17]: Same as other comment ?

**Review by TomDUVAL-MAHE — CHANGES_REQUESTED** [review]

**Review by LucasTesnier — APPROVED** [review]

**Review by TomDUVAL-MAHE — APPROVED** [review]

**Pull request merged**